source/extensions/filters/http/jwt_authn/verifier.cc

```diff
@@ -0,0 +1,228 @@
+      onCompleteHelper(status, context);
+    }
+  }
+
+protected:
+  // The parent group verifier.
+  const BaseVerifierImpl* const parent_;
+};
+
+VerifierPtr innerCreate(const JwtRequirement& requirement,
+                        const Protobuf::Map<ProtobufTypes::String, JwtProvider>& providers,
+                        const AuthFactory& factory, const Extractor& extractor,
+                        const BaseVerifierImpl* parent);
+
+namespace {
+
+// Provider specific verifier
+class ProviderVerifierImpl : public BaseVerifierImpl {
+public:
+  ProviderVerifierImpl(const std::vector<std::string>& audiences, const AuthFactory& factory,
+                       const JwtProvider& provider, const BaseVerifierImpl* parent)
+      : BaseVerifierImpl(parent), audiences_(audiences), auth_factory_(factory),
+        extractor_(
+            Extractor::create(provider.issuer(), provider.from_headers(), provider.from_params())),
+        issuer_(provider.issuer()) {}
+
+  void verify(VerifyContext& context) const override {
+    auto auth = auth_factory_.create(audiences_, absl::optional<std::string>{issuer_}, false);
+    auth->sanitizePayloadHeaders(context.headers());
+    auth->verify(context.headers(), extractor_->extract(context.headers()),
+                 [&](const Status& status) { onComplete(status, context); });
+    if (!context.getResponseData(this).has_responded_) {
+      context.addAuth(std::move(auth));
+    }
+  }
+
+private:
+  const std::vector<std::string> audiences_;
+  const AuthFactory& auth_factory_;
+  const ExtractorConstPtr extractor_;
+  const std::string issuer_;
+};
+
+// Allow missing or failed verifier
+class AllowFailedVerifierImpl : public BaseVerifierImpl {
+public:
+  AllowFailedVerifierImpl(const AuthFactory& factory, const Extractor& extractor, bool allow_failed,
+                          const BaseVerifierImpl* parent)
+      : BaseVerifierImpl(parent), auth_factory_(factory), extractor_(extractor),
+        allow_failed_(allow_failed) {}
+
+  void verify(VerifyContext& context) const override {
+    auto auth = auth_factory_.create({}, absl::nullopt, allow_failed_);
+    auth->sanitizePayloadHeaders(context.headers());
+    auth->verify(context.headers(), extractor_.extract(context.headers()),
+                 [&](const Status& status) { onComplete(status, context); });
+    if (!context.getResponseData(this).has_responded_) {
+      context.addAuth(std::move(auth));
+    }
+  }
+
+private:
+  const AuthFactory& auth_factory_;
+  const Extractor& extractor_;
+  const bool allow_failed_;
+};
+
+// Base verifier for requires all or any.
+class BaseGroupVerifierImpl : public BaseVerifierImpl {
+public:
+  BaseGroupVerifierImpl(const BaseVerifierImpl* parent) : BaseVerifierImpl(parent) {}
+
+  void verify(VerifyContext& context) const override {
+    for (const auto& it : verifiers_) {
+      if (!context.getResponseData(this).has_responded_) {
+        it->verify(context);
```

Comment:
Here we start multiple verification parallelly.   It is good for "require_all" case,  but I think we should not do that for "require_any" case,  it is better to sequentially for "any" case

Reply:
I'm not sure serializing these verifies is a good idea. Best case all JWKs are in cache, and the code here would behave as if it is serialized because is completed would be true. Worst case here is all inner verifies require JWKs to be fetched over the network. Why not greedy fetch all the keys?